api.go

```diff
@@ -363,6 +363,14 @@ type RepoList struct {
 	Minimal map[uint32]*MinimalRepoListEntry
 }
 
+// MuxSearcher is a Sourcegraph addition. It allows to hydrate a Searcher with
+// out-of-band contextual information, like the visibility of the repo.
+type MuxSearcher interface {
```

Comment:
Can you expand on the rationale for the `Mux` prefix? It's not obvious to me how it fits into this change set.

Reply:
The Mux prefix is supposed to say "This is a Searcher with the added capability to mix in information at runtime". Maybe you find a more intuitive name?

Details:
ShardedSearcher holds a collection of shards and is responsible for lock management. If we want to react to changes of the visibility, we need to acquire an exclusive lock (pause all searches) and update the shards (IE indexData). MuxSearcher extends Searcher with the method `setVisibility`. Effectively this gives shardedSearcher a way to mutate indexData while holding the exlusive lock.

Alternatively we could handed new visibiliyt data as an argument to Searcher but this would either require to change the interface Searcher (used everywhere) or extend the SearchOpts (visibility is not really a search option).

Reply:
> Alternatively we could hand new visibility data as an argument to Searcher but this would either require to change the interface Searcher (used everywhere) 

Couldn't we pass the visibility data to the relevant searcher constructors?

Reply:
Yes, that would be an option, but that would mean we only update metadata by reloading the shards. So far our thinking was that it should be possible to mix-in metadata at runtime without reloading or reindexing anything. Maybe it is worth revisiting this dogma.